tensorflow_probability/python/experimental/mcmc/particle_filter.py

```diff
@@ -359,88 +490,122 @@ def particle_filter(observations,
+        parallel_iterations=parallel_iterations,
+        unbiased_gradients=unbiased_gradients,
+        num_timesteps=num_timesteps,
+        particles_dim=particles_dim,
+        trace_fn=trace_fn,
+        loop_seed=loop_seed,
+        never_trace=never_trace
+    )
 
-    # Use `trace_scan` rather than `sample_chain` directly because the latter
-    # would force us to trace the state history (with or without thinning),
-    # which is not always appropriate.
-    def seeded_one_step(seed_state_results, _):
-      seed, state, results = seed_state_results
-      one_step_seed, next_seed = samplers.split_seed(seed)
-      next_state, next_results = kernel.one_step(
-          state, results, seed=one_step_seed)
-      return next_seed, next_state, next_results
+    return traced_results
 
-    final_seed_state_result, traced_results = loop_util.trace_scan(
-        loop_fn=seeded_one_step,
-        initial_state=(loop_seed,
-                       initial_weighted_particles,
-                       kernel.bootstrap_results(initial_weighted_particles)),
-        elems=tf.ones([num_timesteps]),
-        trace_fn=lambda seed_state_results: trace_fn(*seed_state_results[1:]),
-        trace_criterion_fn=(
-            lambda seed_state_results: trace_criterion_fn(  # pylint: disable=g-long-lambda
-                *seed_state_results[1:])),
-        static_trace_allocation_size=static_trace_allocation_size,
-        parallel_iterations=parallel_iterations)
 
-    if trace_criterion_fn is never_trace:
-      # Return results from just the final step.
-      traced_results = trace_fn(*final_seed_state_result[1:])
+def sample_at_dim(initial_state_prior, dim, num_samples, seed=None):
+  if type(initial_state_prior.batch_shape) is dict:
+    model_dict = initial_state_prior.model
+    sampled_model = {}
 
-    return traced_results
+    for key in model_dict.keys():
+      d = model_dict[key]
+      batch_shape = d.batch_shape
+      d = batch_reshape.BatchReshape(d, batch_shape[:dim] + [1] + batch_shape[dim:])
+      d = batch_broadcast.BatchBroadcast(d, batch_shape[:dim] + [num_samples] + batch_shape[dim:])
+      sampled_model[key] = d.sample(seed=seed)
+
+    return sampled_model
+
+  else:
+    batch_shape = initial_state_prior.batch_shape
+    initial_state_prior = batch_reshape.BatchReshape(initial_state_prior, batch_shape[:dim] + [1] + batch_shape[dim:])
+    initial_state_prior = batch_broadcast.BatchBroadcast(initial_state_prior, batch_shape[:dim] + [num_samples] + batch_shape[dim:])
+    return initial_state_prior.sample(seed=seed)
 
 
 def _particle_filter_initial_weighted_particles(observations,
                                                 observation_fn,
                                                 initial_state_prior,
                                                 initial_state_proposal,
-                                                num_particles,
+                                                num_inner_particles,
```

Comment:
I think this argument should still be called `num_particles`.

Callers of this function might have different particle dimensions for different nested SMCs / particle filters -- e.g., "inner" and "outer".  But from the perspective of this function, I don't think the concept of "inner particles" makes sense -- we're just creating the initial state for some particle filter that has `num_particles` particles, indexed by dimension `particles_dim`.